### PR TITLE
Adds Batching Proposal

### PR DESCRIPTION
**Pull request by michaelstaib**
Within the Composite Schema WG we have discussed a new batching format that is in the first place meant for Subgraphs/Source Schema in a federated graph. This new variable batching allows the distributed GraphQL executor to execute the same operation with a set of variables.

https://github.com/graphql/composite-schemas-spec/issues/25

### CODE REVIEW

**Comment by Shane32** [comment]
Does the distributed GraphQL executor currently support 'regular' batching requests?

**Comment by michaelstaib** [comment]
@Shane32 we will specify this as well ... we call it at the moment request batching ... the idea is that a request batch can also consist of variable batches.

**Comment by Shane32** [comment]
So it would support variable batching within request batching then?

**Comment by michaelstaib** [comment]
Yes, this is the current discussion. There are a lot of constraints we will put in place for the first iteration of this, we have explored this also in combination with subscriptions and all. But for this initial appendix we are focusing of variable batching first as this will be the minimum requirement for the composite schema spec.

**Comment by Shane32** [comment]
Ok.  I would suggest that we do not use the jsonl format.

1. If there is variable batching within request batching, there is no clear way to format such a response.  Obviously wrapping a jsonl response in a JSON list produces unparsable JSON due to the missing commas:

```
[
{"data":{"hello":"world-1"}}
{"data":{"hello":"world-2"}}
,
{"data":{"name":"john-1"}}
{"data":{"name":"john-2"}}
]
```

Perhaps it could be specified that when request batching was layered on top of variable batching, then the lists are flattened.  I'm not sure this is the best approach, but it's feasible.

2. If this scenario is not supported, implying that the request is request batching OR variable batching, then there is no reason to differentiate the response formats.

3. Parsing jsonl is not common in JSON environments, at least not the ones I'm familiar with.  Attempting to parse such a response within .NET will throw an exception upon encountering the next line in the jsonl response.  Similarly the `JSON.parse` method of javascript does not support jsonl and will throw.  In any of these environments, one would have to code another layer to separate the responses before decoding the results, adding code complexity in addition to having a likely-slower implementation.  It seems much cleaner and easier to parse the JSON and then iterate through the response list.

Let's use what is commonplace now and supported by various servers and clients alike.

Perhaps as a separate appendix, the jsonl format is described as an optional response format for batching requests (request batching or variable batching).  There it can state that if multiple batching approaches are used, the lists are flattened.  But I just don't see the benefit of adding another response format.

**Comment by michaelstaib** [comment]
There actually is we have specified that there is a requestIndex and a variableIndex in the response structure ... we do need these also as there is defer and stream involved. Every execution could yield to a stream of responses. We have this also already implemented with reference implementations ... I will get through the spec text this week and then you will see how this plays out.

**Comment by michaelstaib** [comment]
BTW ... we will introduce `requestIndex` and `variableIndex` also since the server should be able reorder responses and not make the consumer wait just in order to return completed results in order.

**Comment by michaelstaib** [comment]
@Shane32 I have put a bit more about the response in.

File: spec/Appendix B -- Variable Batching.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+## B. Appendix: Variable Batching
```
Comment (michaelstaib) [inline spec/Appendix B -- Variable Batching.md:1]: After discussion in the composite schema working group we will have a single appendix about batching as this would allow for better example that show also both request batching and variable batching in combination.

Comment (michaelstaib) [reply spec/Appendix B -- Variable Batching.md:1]: @benjie shall we do also some example in here with defer and stream ... I asked that because defer and stream is not yet handled in the spec.

Comment (benjie) [reply spec/Appendix B -- Variable Batching.md:1]: For now: no; but would be good to have the text handy anyway.

File: spec/Appendix B -- Variable Batching.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  Document to execute.
+- **{variables}** (Required, array of maps):
+  Instead of a single map, the `variables` parameter is an array of maps. Each map in the array represents a different set of variables to be used in executing the operation.
+- {extensions} - (_Optional_, map): This entry is reserved for implementors to
+  extend the protocol however they see fit.
+
+### Accept
+
+A client SHOULD indicate the media types that it supports in responses using the `Accept` HTTP header as specified in [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231).
+
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:56]: No need to repeat this.

```suggestion
```

File: spec/Appendix B -- Variable Batching.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+For **variable batching requests**, the client SHOULD include the media type `application/graphql+jsonl` in the `Accept` header to indicate that it expects a batched response in JSON Lines format.
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:57]: There's a few issues with this:

`+jsonl` isn't allowed; see: https://www.rfc-editor.org/rfc/rfc6838.html#section-4.2.8

> The primary guideline for whether a structured type name suffix is
> registrable is that it be described by a readily available
> description, preferably within a document published by an established
> standards-related organization, and for which there's a reference
> that can be used in a Normative References section of an RFC.

`application/graphql` is incorrect (that's normally used for GraphQL documents, not GraphQL responses) - I think you meant `application/graphql-response`?

This isn't _a_ response but instead a number of responses.

Suggestion: just use `application/jsonl` as mentioned by https://jsonlines.org/ - this is sufficient to solve the issues that `application/json` has w.r.t. status codes/etc.

File: spec/Appendix B -- Variable Batching.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+If the client supplies an Accept header, the client SHOULD include the media type `application/graphql+jsonl` in the Accept header.
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:59]: I don't see the value in this sentence given the previous one - am I missing something?

```suggestion
```

File: spec/Appendix B -- Variable Batching.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+If the client does not supply an `Accept` header, the server MAY respond with a default content type of `application/graphql+jsonl`, or it may use any other media type it supports. However, to ensure compatibility and clarity, it is RECOMMENDED that the client explicitly states its preferred media types using the `Accept` header.
+
+### POST
+
+A **variable batching request** instructs the server to perform a query or mutation operation multiple times, once for each set of variables provided. The request MUST have a body that contains the values of the _variable batching request_ parameters encoded in the `application/graphql+jsonl` media type, or another media type supported by the server.
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:65]: We wouldn't encode the parameters in JSONL? Just a simple JSON value is fine - they're all sent at the same time.

File: spec/Appendix B -- Variable Batching.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+A client MUST indicate the media type of a request body using the `Content-Type` header as specified in [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231).
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:67]: Already part of the main spec, I don't think we need to repeat it?

```suggestion
```

File: spec/Appendix B -- Variable Batching.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    {
+      "id": "QVBJcy5ndXJ3"
+    }
+  ]
+}
+```
+
+### Response
+
+When a server receives a well-formed _variable batching request_, it MUST return a well‐formed stream of _GraphQL responses_. Each response in the stream corresponds to the result of validating and executing the requested operation with one set of variables. The server's response stream describes the outcome of each operation, including any errors encountered during the execution of the request.
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:143]: I don't think we should use the word "stream" here. I think "list" is fine; I think the default media type would likely be e.g. `application/graphql-response-list+json` (a JSON array of GraphQL responses) but `application/jsonl` would allow for easier early parsing of responses when the client is stream-capable.

File: spec/Appendix B -- Variable Batching.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+A server must comply with [RFC7231](https://datatracker.ietf.org/doc/html/rfc7231).
+
+Each response in the stream follows the standard GraphQL response format, with the addition of a required `variableIndex` field at the top level of each response. The `variableIndex` indicates the index of the variables map from the original request, allowing the client to associate each response with the correct set of variables.
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:147]: The definition of the `application/graphql-response` type should be modified such that other keys are allowed iff they are specified as part of a specification or appendix hosted under the `graphql.org` domain; then this spec can "extend" that type with these keys. May want to go a slightly different direction on this in future.

File: spec/Appendix B -- Variable Batching.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+**Note:** The server MAY respond with results in any order. The `variableIndex` field ensures that clients can correctly match each response to its corresponding set of variables, regardless of the order in which the responses are returned.
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:149]: This isn't non-normative, so shouldn't be a Note:

```suggestion
The server MAY respond with results in any order. The `variableIndex` field ensures that clients can correctly match each response to its corresponding set of variables, regardless of the order in which the responses are returned.
```

Also, spec-md expect's notes to not be bolded.

File: spec/Appendix B -- Variable Batching.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+
+#### Response Structure
+
+Each line in the JSON Lines (JSONL) response MUST include the following fields:
```
Comment (benjie) [inline spec/Appendix B -- Variable Batching.md:153]: The combination of `MUST` and `optional` is confusing. I would change this to something more along the lines of "in addition to the regular fields expected on a GraphQL response, each entry in a variable batching response MUST include the `variableIndex` field..."

**Review by benjie — CHANGES_REQUESTED** [review]
We should remove some of the repetition of the core spec in this appendix, try and see it more as "additions" (extra fields, extra rules, changes in behavior) rather than a re-implementation.

**Comment by benjie** [comment]
> 3. Parsing jsonl is not common in JSON environments, at least not the ones I'm familiar with.  Attempting to parse such a response within .NET will throw an exception upon encountering the next line in the jsonl response.  Similarly the `JSON.parse` method of javascript does not support jsonl and will throw.  In any of these environments, one would have to code another layer to separate the responses before decoding the results, adding code complexity in addition to having a likely-slower implementation.  It seems much cleaner and easier to parse the JSON and then iterate through the response list.

It's not really fair to expect a JSON parser to parse JSONL... it's a different format. However; to make JSONL parseable as JSON is very straightforward; here's a JSONL parser for JS:

```js
const parseJSONL = (jsonl) => JSON.parse(`[${jsonl.trim().replaceAll('\n', ',')}]`);
```

I'm not sure that you can really say that this significantly "increases the code complexity"? I would expect .NET to also be capable of:

- trimming whitespace (including newlines) from both ends of a string
- replacing all instances of a newline in the string with a comma
- prepending an `[` and appending a `]`
- parsing the result as JSON

JSONL is an incredibly straightforward format. Most importantly for this use case, it allows you to process the values as they stream down because you just scan through it looking for the next `\n` and when you hit it everything up until that point is parseable as JSON - you don't need a streaming parser as you would if you were to use JSON as the data type.

All that said, I think we should RECOMMEND that clients and servers implement JSONL, but we should REQUIRE that servers support plain JSON arrays. Which I think is already the case in this text (since SHOULD is equivalent to RECOMMEND according to RFC2119).

**Comment by Shane32** [comment]
> here's a JSONL parser for JS:
>
> ```js
> const parseJSONL = (jsonl) => JSON.parse(`[${jsonl.trim().replaceAll('\n', ',')}]`);
> ```

Fair enough; I didn't think of that.

> I'm not sure that you can really say that this significantly "increases the code complexity"? I would expect .NET to also be capable of:

True, .NET could perform string manipulation also fairly easily.  However, the typical .NET JSON parser performs on a UTF8 byte stream so that there is no additional memory allocations beyond the minimum required to deserialize the data.  Converting the incoming data to a string, performing a number of string manipulations on it, and then parsing it would considerably slow down the JSON engine.  To maintain speed on par with the default implementation, you'd have to write a specialized wrapper that parsed the characters as they were being read, looking for `\n` and either (a) reading each JSON line as it is streamed, or (b) creating a virtual list of 'sub-streams' for further parsing. Or perhaps a streaming wrapper that changed the characters similar to the JS conversion above so it could be read as JSON.

> JSONL is an incredibly straightforward format. Most importantly for this use case, it allows you to process the values as they stream down because you just scan through it looking for the next `\n` and when you hit it everything up until that point is parseable as JSON - you don't need a streaming parser as you would if you were to use JSON as the data type.

Probably true.  I may have another viewpoint if I was more familiar with the use case (which I am not).

**Comment by benjie** [comment]
> Probably true. I may have another viewpoint if I was more familiar with the use case (which I am not).

In general batching is used when the client has a lot of queries to execute all at the same time (e.g. as the result of rendering a React tree). In traditional batching, the server receives this as a list, executes them in parallel, and then returns the resulting array - the result being that none of the components can render until the slowest of all the queries has finished executing. By allowing a) the server to return the results in order of execution completion (rather than in request order) and b) the client to determine easily when a result is ready (e.g. by scanning the response for a `\n` character) the client can _choose_ to read and render results from faster queries whilst slower queries are still executing, and this can be achieved without needing to use an alternative protocol such as SSE or websockets. If the client can't be bothered to do this, then nothing changes and they just read and process the final JSONL result in much the same way they would with JSON.

**Comment by benjie** [comment]
> Or perhaps a streaming wrapper that changed the characters similar to the JS conversion above so it could be read as JSON (but that would be even more complex).

Would it? You can initialize your binary UTF8 data buffer with 0x5b (`[`) (rather than being empty as you normally would), read the JSONL from the network and write into the buffer, then finally write a closing 0x5d (`]`) to the buffer. Once complete, just before you would parse the JSON, first check if the penultimate byte is 0x0a (`\n`) and if so overwrite it with 0x20 (` `) to remove the trailing newline (if any); then scan over this entire buffer and overwrite each 0x0a (`\n`) with a 0x2c (`,`). That's it - there's no resizing the buffer, it's safe to perform against UTF8 (even though it's a multi-byte format) because we're in the ASCII plane, it's just a simple prepend, append, byte-wise find and replace, and then using your regular JSON parser. The performance impact should be minimal; but you could speed it up with SIMD if you really wanted to.

**Comment by Shane32** [comment]
> > Probably true. I may have another viewpoint if I was more familiar with the use case (which I am not).
> 
> In general batching is used when the client has a lot of queries to execute all at the same time (e.g. as the result of rendering a React tree). In traditional batching, the server receives this as a list, executes them in parallel, and then returns the resulting array - the result being that none of the components can render until the slowest of all the queries has finished executing. By allowing a) the server to return the results in order of execution completion (rather than in request order) and b) the client to determine easily when a result is ready (e.g. by scanning the response for a `\n` character) the client can _choose_ to read and render results from faster queries whilst slower queries are still executing, and this can be achieved without needing to use an alternative protocol such as SSE or websockets. If the client can't be bothered to do this, then nothing changes and they just read and process the final JSONL result in much the same way they would with JSON.

I see.  Why does this PR focus on allowing any return order for variable batching vs regular batching then?  If anything, variable batching would be much more likely to have a consistent execution time across each request and so having the responses return in any order is much less useful.  Maybe we should focus on a flag of some sort to allow this behavior for traditional batching requests, and then take the solution to this PR for consistency.

**Comment by Shane32** [comment]
> > Or perhaps a streaming wrapper that changed the characters similar to the JS conversion above so it could be read as JSON (but that would be even more complex).
> 
> Would it? You can initialize your binary UTF8 data buffer with 0x5b (`[`) (rather than being empty as you normally would), read the JSONL from the network and write into the buffer, then finally write a closing 0x5d (`]`) to the buffer. Once complete, just before you would parse the JSON, first check if the penultimate byte is 0x0a (`\n`) and if so overwrite it with 0x20 (` `) to remove the trailing newline (if any); then scan over this entire buffer and overwrite each 0x0a (`\n`) with a 0x2c (`,`). That's it - there's no resizing the buffer, it's safe to perform against UTF8 (even though it's a multi-byte format) because we're in the ASCII plane, it's just a simple prepend, append, byte-wise find and replace, and then using your regular JSON parser. The performance impact should be minimal; but you could speed it up with SIMD if you really wanted to.

I meant a streaming wrapper.  One that does not read the entire result into memory at any time.  Yes, I believe it would be more complex than the other two options I presented.  For instance, if you read a blob of data and the last character is a LF, you can't pass it to the JSON parser but instead must cache it until the next read so you know if this is the EOF and it must be overwritten with `]` or if there is more data and it must be `,`.  It just would get sorta complex.  And it would require you to read the JSON in a streaming fashion too, which would be even more complex, if you wanted to read the results as they were streamed in.

I'm okay with the complexity if it's optional and has benefits, mind you.  I would agree JSONL seems to be a better format for reading streaming responses, either in .NET or JS, where you want to read individual GraphQL responses as they are transmitted.

**Comment by Shane32** [comment]
> Maybe we should focus on a flag of some sort to allow this behavior for traditional batching requests, and then take the solution to this PR for consistency.

For example, maybe anytime a JSONL format is requested for a batching request (traditional or variable batching), a `batchIndex` is added and results are returned in any order.

**Review by martinbonnin — APPROVED** [review]
One nit: the name of the appendix file should be just `Batching`. LGTM otherwise.

File: spec/GraphQLOverHTTP.md [hunk 11]
```diff
@@ -432,6 +432,15 @@ A server must comply with
 The body of the server's response MUST follow the requirements for a
 [GraphQL response](#sec-Response), encoded directly in the chosen media type.
 
+A server MUST NOT include additional top-level entries in a _GraphQL response_
+unless those entries are explicitly defined by one of:
+
+1. The GraphQL specification.
+2. This specification, including its appendices.
+3. Another specification or appendix published under the `graphql.org` domain.
```
Comment (benjie) [inline spec/GraphQLOverHTTP.md:440]: Item 3 implies GAPs can extend it. I don't think that's wise - GAPs should use `extensions`. Is this specifically so the composite schemas spec can extend it? What are you thinking about here?

File: spec/Appendix B -- Batching.md [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+
+Variable batching and request batching MAY be used independently or together.
+
+Note: GraphQL subscriptions are beyond the scope of this specification, so this
+appendix defines batching semantics for query and mutation operations only.
+
+### Motivation
+
+Field-level batching can solve some cases, but it does not cover all shapes of
+variable-dependent queries. For example:
```
Comment (benjie) [inline spec/Appendix B -- Batching.md:18]: "field-level batching" is not well defined.

File: spec/Appendix B -- Batching.md [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        id
+        name
+      }
+    }
+  }
+}
+```
+
+A list of variable maps allows this operation to execute multiple times in one
+HTTP request, each time with a different pair of variable values.
```
Comment (benjie) [inline spec/Appendix B -- Batching.md]: I didn't find this wording obvious, but I'm not a big fan of my suggested replacement either:

```suggestion
Supplying the document to execute alongside a _list_ of {variableValues} maps
allows this operation to execute multiple times in a single HTTP request, each
time with a different {variableValues} map.
```

File: spec/Appendix B -- Batching.md [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+
+### Request Shapes
+
+A batching request is sent as a POST request with a JSON-encoded body.
+
+This appendix extends the `{variables}` request parameter:
+
+- `{variables}` MAY be a map (as defined in the main specification).
+- `{variables}` MAY be a list of maps for variable batching.
+- If `{variables}` is a list, each item in the list MUST be a map.
```
Comment (benjie) [inline spec/Appendix B -- Batching.md]:
```suggestion
To add support for variable batching, the {variables} _request parameter_ is extended such that it may now also accept a list of maps.
```

File: spec/Appendix B -- Batching.md [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+A batching request body MUST be one of the following:
+
+- A single _GraphQL-over-HTTP request_ object.
+- A list of _GraphQL-over-HTTP request_ objects.
```
Comment (benjie) [inline spec/Appendix B -- Batching.md]:
```suggestion
To add support for request batching, the JSON encoding of a _GraphQL-over-HTTP
request_ is extended such that the body of the request may also be a list of _GraphQL-over-HTTP request_ maps.
```

File: spec/Appendix B -- Batching.md [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+
+If a request object's `{variables}` value is a list, the server MUST execute
+that request once per variables entry.
+
+If the request body is a list, each entry in that list MUST be processed as an
+independent _GraphQL-over-HTTP request_.
```
Comment (benjie) [inline spec/Appendix B -- Batching.md:54]: We need to encourage them to discard requests where the total number of requests to execute is over some threshold. But also, I think this is more part of execution behavior than request shape.

```suggestion
```

File: spec/Appendix B -- Batching.md [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+
+### Execution Semantics
+
+When a request object's `{variables}` value is a list, each variables map
+represents a distinct execution of that request object.
+
+The server MAY execute these executions in any order, including concurrently,
+and the client MUST NOT assume serial execution or any ordering based on
+variables list position.
```
Comment (benjie) [inline spec/Appendix B -- Batching.md:62]:
```suggestion
When a request object's `{variables}` value is a list, each variables map
represents a distinct execution of that request object.
The server MAY execute these executions in any order, including concurrently,
and the client MUST NOT assume serial execution or any ordering based on
variables list position.
```

**Review by benjie — CHANGES_REQUESTED** [review]
The more I look at this, the more I feel it should be baked into the main spec. Variable batching at least... I could happily do without operation batching (in fact, I'd argue for us to avoid specifying it).

I think we might need to abstract the spec a little to make way for Trusted Docs and batching

Like:

```mermaid
flowchart TD
  HTTPReq --> GraphQLHTTPRequest
  GraphQLHTTPRequest --> GraphQLRequest
  GraphQLRequest --> Execute
  Execute --> GraphQLResponse
  GraphQLResponse --> GraphQLHTTPResponse
  GraphQLHTTPResponse --> HTTPResponse
```

Then we can do:

```mermaid
flowchart TD
  HTTPReq --> TrustedDocRequest["TrustedDocsRequest {id:'abc', variables: [{...}, ...]}"]
  TrustedDocRequest --> GraphQLHTTPRequest["GraphQLHTTPRequest {query:'...', variables: [{...}, ...]}"]

  GraphQLHTTPRequest --> GraphQLReq1["GraphQLRequest {query:'...', variables: {...}}"]
  GraphQLReq1 --> Execute1["Execute"]
  Execute1 --> GraphQLResponse1["GraphQLResponse"]
  GraphQLResponse1 --> GraphQLHTTPResponse

  GraphQLHTTPRequest --> GraphQLReq2["GraphQLRequest {query:'...', variables: {...}}"]
  GraphQLReq2 --> Execute2["Execute"]
  Execute2 --> GraphQLResponse2["GraphQLResponse"]
  GraphQLResponse2 --> GraphQLHTTPResponse

  GraphQLHTTPRequest --> GraphQLReq3["GraphQLRequest {query:'...', variables: {...}}"]
  GraphQLReq3 --> Execute3["Execute"]
  Execute3 --> GraphQLResponse3["GraphQLResponse"]
  GraphQLResponse3 --> GraphQLHTTPResponse

  GraphQLHTTPResponse --> HTTPResponse["HTTPResponse 207 Multi-Status"]
```

File: spec/Appendix B -- Batching.md (hunk 16 above)
Comment (michaelstaib) [reply spec/Appendix B -- Batching.md:54]: If the server is allowed to discard requests then this must be specified in some way. Like it cannot simply discard requests in the middle without any notice. Otherwise we need another marker to pin which requests have to be reissued. I a server like executes 20 requests and then says limit reached a stop now that could be handled... although to be honest I would find it better if the server would simply reject it all and say i can only process 20 requests at a time so i reject your batch.

This ties btw back in with capabilities. In the gateway case the gateway could inspect the capabilities and send exactly the correct amount of requests in a batch.

File: spec/GraphQLOverHTTP.md (hunk 11 above)
Comment (michaelstaib) [reply spec/GraphQLOverHTTP.md:440]: My thinking was exactly that there might be transport specific spec bound to something like the composite schema spec. But I do not have anything in mind yet, just thought this gives flexibility. But i maybe not needed.

I was not thinking about GAPs at the time.

**Comment by michaelstaib** [comment]
@benjie regarding request batching ... we need to spec it, like for Gateways the ability to pack two different requests to the same subgraph really saves on pressure and improves perf. In apollo federation this is solved with alias batching as they can pass in untyped data into the representation they can basically do the same as combining variable batching and request batching combined.

The pair does work with trusted documents. I would agree to moving variable batching into the main spec and leaving request batching as an appendix. I would not drop it.

File: spec/GraphQLOverHTTP.md (hunk 11 above)
Comment (benjie) [reply spec/GraphQLOverHTTP.md:440]: Let's remove it and add it back when needed.

File: spec/Appendix B -- Batching.md (hunk 16 above)
Comment (benjie) [reply spec/Appendix B -- Batching.md:54]: Yes, I think all or none - it should decide up front if it's attempting them or not.

That said, mutations would be serialized and may hit timeouts, and the side effects might not be able to be undone, so... :grimacing:  I guess just execute them all and the later ones produce errors...

... We really need to add some standard error codes to GraphQL errors.

**Comment by benjie** [comment]
I think there's enough evidence out there in the ecosystem that GraphQL needs variable batching _independent_ of Federation that I think it's not unreasonable to include it as a recommended feature. We can indeed use service capabilities to indicate not only which types of batching are supported but also what their limits should be (and whether or not they can be combined).

**Comment by bclozel** [comment]
Hello!

I'm a maintainer of the [spring-graphql](https://github.com/spring-projects/spring-graphql) project. I looked at this appendix from the angle of "what would it take to support this in a statically-typed JVM HTTP transport layer," and ran into a few questions/concerns I think are worth raising.

### Mutation operations and request batching

The appendix currently scopes batching to "query and mutation operations only," excluding subscriptions. This is the only sentence in the whole appendix that addresses operation types. I couldn't find any reference to mutations in the motivating composite-schemas-spec issue (graphql/composite-schemas-spec#25).

That matters because the appendix gives no ordering guarantee for either batching mode:

> Variable batching: "The server MAY execute these executions in any order, including concurrently, and the client MUST NOT assume serial execution or any ordering based on variables list position."

> Request batching: "the server MAY process those request objects in any order, including concurrently, and the client MUST NOT assume ordering based on request list position."

For queries this is harmless, but I suspect that for mutations it isn't. Tiday, clients can put multiple mutation-typed root fields in a single document, and the base GraphQL spec already requires serial, in-order execution. I'd like to understand whether the WG intends for mutation batching to carry the same no-ordering guarantee as query batching, or whether the spec should only consider queries in the for now.

### Why support request batching and variable batching combined?

I'd like to understand the concrete use case for allowing both mechanisms to compose in the same HTTP request. The appendix treats them as fully orthogonal and combinable, which produces a two-dimensional indexing scheme (`requestIndex` × `variableIndex`) and multiplies the parsing, response-correlation, and ordering-semantics surface that implementers need to build and test.

Looking at composite-schemas-spec#25, which is cited as the motivation for this appendix, request batching and variable batching were presented as alternative solutions to the same single problem with distinct trade-offs. Could you elaborate on why a combination is supported here?

### No dedicated media type for either batching mode

Request batching adds complexity to parsing `application/json` request bodies: the top-level body can be a single request object or a list of them, and independently, each request object's `variables` value can be a single map or a list of maps. None of this is signaled by `Content-Type`. I suspect that servers have to inspect the body's shape before it knows which of up to four shapes it's holding, and that determination has to happen *before* any typed deserialization can occur.

I understand that this is currently being used out there so I'm curious. How do typed languages approach this problem? Doesn't this create performance and implementation challenges? Are frameworks doing a raw/tree-based JSON pre-parse or have they moved away from typed deserialization entirely for the request path? For a framework like ours, built around a single typed deserializer at the transport boundary, this forces either a two-pass parse or pushing `Object`-typed (Map or List) fields deep into the request model and interceptor pipeline, which we'd rather avoid if there's a cleaner option.

Given that this spec involves quite a few HTTP concepts already (response status, response content type, HTTP verbs behavior), I'm wondering why this proposal is not considering this angle. Is it because of the cost of defining new media types in general?

Thanks!

**Comment by martinbonnin** [comment]
> Given that this spec involves quite a few HTTP concepts already (response status, response content type, HTTP verbs behavior), I'm wondering why this proposal is not considering this angle. Is it because of the cost of defining new media types in general?

@bclozel we discussed adding new media types in the [last working group](https://www.youtube.com/watch?v=dSPxCFxgeR0). I think this is generally the direction this group is taking. It requires some prep work (see also https://github.com/graphql/graphql-over-http/pull/426 and https://github.com/graphql/graphql-over-http/pull/427) but we'll get there!

There is a working group tomorrow, please feel free to join by adding your name to [the agenda](https://github.com/graphql/graphql-over-http/blob/main/working-group/agendas/2026/08-Aug/27-graphql-over-http-wg-august-2026.md).

**Comment by dariuszkuc** [comment]
> Why support request batching and variable batching combined?

TLDR it is an optimization.

When doing the subsequent entity fetches, you may have to `@lookup` multiple entities in a single fetch - if they are the same entity you would use _variable_ batching (as it is the same query and you only need to send different args), if they are different entities you would need to rely on _request_ batching (as different entities would have different `@lookup` queries) and you would use _variable_ + _request_ batching if you have a combination of both then you would.
